addons/stock_account/models/stock.py

```diff
@@ -87,83 +88,211 @@ def _get_price_unit(self):
             move = self.search([('product_id', '=', self.product_id.id),
                          ('state', '=', 'done'), 
                          ('location_id.usage', '=', 'internal'), 
-                         ('location_dest_id.usage', '!=', 'internal')], order='date desc', limit=1)
+                         ('location_dest_id.usage', '!=', 'internal'),
+                         ('company_id', '=', self.company_id.id)], order='date desc', limit=1)
             if move:
                 return move.price_unit or self.product_id.standard_price
         return self.product_id.standard_price
 
     def _update_future_cumulated_value(self, value):
         self.ensure_one()
         moves = self.search([('state', '=', 'done'), 
-                     ('date', '>',  self.date), 
-                     ('product_id', '=', self.product_id.id)])
+                     ('date', '>=',  self.date),
+                     ('id', '>', self.id), 
+                     ('product_id', '=', self.product_id.id), 
+                     ('company_id', '=', self.company_id.id),
+                     '|', '&', ('location_id.usage', 'in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'not in', ('internal', 'transit')), 
+                    '&', ('location_id.usage', 'not in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'in', ('internal', 'transit'))])
         for move in moves:
-            move.value += value
+            move.cumulated_value += value
+
+    def _get_in_domain(self):
+        return [('product_id', '=', self.product_id.id), 
+                ('state', '=', 'done'), 
+                ('company_id', '=', self.company_id.id), 
+                ('location_id.usage', 'not in', ('internal', 'transit')), 
+                ('location_dest_id.usage', 'in', ('internal', 'transit'))]
+
+    def _get_out_domain(self):
+        return [('product_id', '=', self.product_id.id), 
+                ('state', '=', 'done'), 
+                ('company_id', '=', self.company_id.id), 
+                ('location_id.usage', 'in', ('internal', 'transit')), 
+                ('location_dest_id.usage', 'not in', ('internal', 'transit'))]
+
+    def _get_all_domain(self):
+        return [('product_id', '=', self.product_id.id), 
+                 ('state', '=', 'done'), 
+                 ('company_id', '=', self.company_id.id),
+                 '|', '&', ('location_id.usage', 'in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'not in', ('internal', 'transit')), 
+                    '&', ('location_id.usage', 'not in', ('internal', 'transit')), 
+                    ('location_dest_id.usage', 'in', ('internal', 'transit'))]
+        
+    def _is_in_move(self):
+        return self.location_id.usage not in ('internal', 'transit') and self.location_dest_id.usage in ('internal', 'transit')
```

Comment:
I know what you mean by `If it is within the company` because I and my colleagues hit this same issue a long time ago, the thing with merchandise going into transit, is that it creates Journal Entries when going out and coming in, if that is the case we have to be consistent with the value of the merchandise (if creating JE), or we have to be consider that merchandise going through inter-transit/same-company does not create JEs, and as you said, it is complex issue.